File: leetcode/python/0101_symmetric_tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        def helper(l, r):
            if l == None and r != None: return False
            if l != None and r == None: return False
            if l != None and r != None:
                if not helper(l.left, r.right): return False
                if l.val != r.val: return False
                if not helper(l.right, r.left): return False
            return True

        if not helper(root.left, root.right): return False
        return True
        

File: leetcode/java/0101_symmetric_tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean helper(TreeNode l, TreeNode r) {
        if (l != null && r != null) {
            if (helper(l.left, r.right) == false) return false;
            if (l.val != r.val) return false;
            if (helper(l.right, r.left) == false) return false;
        }

        if (l == null && r != null) return false;
        if (l != null && r == null) return false;

        return true;
    }

    public boolean isSymmetric(TreeNode root) {
        return helper(root.left, root.right);
    }
}